### DISCUSSION TITLE
driving vimium c with selenium?

### DISCUSSION THREAD

**Discussion by smallstepman** [gdh1995/vimium-c#492] [Ideas]
Hi! Love the extension, been using it for quite a while now, it has been real time saver in my workflow, thanks!

I'm having an exotic idea in mind, which is to try to drive chrome with emacs or WM (i3 in my case). To put in in other words, I'd like to use chrome, without actually touching the window itself - 90% of actions (clicking on links, scrolling, coping/inserting text) would be done by selenium, and I'd build necessary scripts and shortcuts to send commands to selenium. This is very similar in spirit to what `nyxt` browser offers, but with limited scope of functionality.

I think it theoretically could be done if I were to combine vimium c together with selenium, tho I'm not yet sure how to do it. I'll attempt to sequence out the steps necessary to achieve it, however the list will not be complete because I miss an important link :link:, which is how to talk to vimium from running selenium engine/chromedriver - I'm not very well versed on how chrome is put together so it's hard to me to figure how to hack around it... I'm hoping someone will be able to help me out here.

#### Steps:
- build script which will listen for IPC calls, and pass those calls to selenium engine
- create set of keybindings for emacs/WM(/AHK for windows user), which will send IPC call to the script from previous step
- (missing part :link:) use selenium to execute vimium c commands

do you think it would be possible to achieve with `Vimium inner URLs` ?

**Comment by gdh1995** [comment]
Do you mean to send commands to Vimium C and let it do concrete tasks?

Then what's the purpose of importing selenium?

I mean, since most features of Vimium C can be accessed from just a keyboard, you may **directly send WM-level keyboard events** to a Chrome window, and then Vimium C will hook them and trigger its commands.

===============

Now Vimium C can recognize and run a big command tree from a bookmark of `vimium://run/...` ( https://github.com/gdh1995/vimium-c/wiki/Vimium-inner-URLs , https://github.com/gdh1995/vimium-c/wiki/Auto-run-a-tree-of-commands ).

However, such URLs can only work when you use Vimium C's commands like `openUrl` and `openBookmark` to open them, but not when you click a bookmark item on the bookmark panel directly.

=============

As for the feasibility, some commands might require your browser is "focused", so I'm not sure whether it's possible or not to send shortcuts from a terminal.

**Reply by smallstepman** [reply]
> Do you mean to send commands to Vimium C and let it do concrete tasks?

Exactly. 

> Then what's the purpose of importing selenium?
> I mean, since most features of Vimium C can be accessed from just a keyboard, you may directly send WM-level keyboard events to a Chrome window, and then Vimium C will hook them and trigger its commands.

My goal here is to never have to focus the Chrome window, that's why I thought about Selenium - a man in the middle which focuses ("drives") the window for me.

**Reply by smallstepman** [reply]
Another way to accomplish this, would be to expose Vimium C via RPC or websocket, allowing the user to connect to it from outside of the browser, however since you've mentioned some commands require the browser window to be focused, I'm also not sure about feasibility

**Reply by smallstepman** [reply]
But considering further the "rpc/websocket" feature, I think the issue with window focus is not significant one, after all, it's a trivial task to write a script which focuses browser's window for   the duration of the time when RPC/websocket call is being made (short duration of time; sub-second)

**Reply by gdh1995** [reply]
Vimium C allow messages from other extensions and it can be used to call any command:
https://github.com/gdh1995/vimium-c/blob/6d286109cc703e350c84302895083ec5beff4bfa/background/main.ts#L101-L110

So if you want you may implement another extension with "native module" to create a websocket server to listen for and forward messages.

But I still think it's enough to send simulated WM keyboard events to a Chrome window.

**Reply by gdh1995** [reply]
Native messaging of chrome extensions: https://developer.chrome.com/docs/apps/nativeMessaging/

**Reply by smallstepman** [reply]
Awesome, many thanks for resources!

**Comment by gdh1995** [comment]
And, since this feature is so far away from usages of a normal "web-extension", I won't write much code for this idea.
